qiskit/primitives/statevector_sampler.py

```diff
@@ -0,0 +1,215 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023, 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Statevector Sampler class
+"""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Iterable
+
+import numpy as np
+from numpy.typing import NDArray
+
+from qiskit import ClassicalRegister, QiskitError, QuantumCircuit
+from qiskit.circuit import ControlFlowOp
+from qiskit.quantum_info import Statevector
+
+from .base import BaseSamplerV2
+from .base.validation import _has_measure
+from .containers import (
+    BitArray,
+    PrimitiveResult,
+    PubResult,
+    SamplerPub,
+    SamplerPubLike,
+    make_data_bin,
+)
+from .containers.bit_array import _min_num_bytes
+from .primitive_job import PrimitiveJob
+from .utils import bound_circuit_to_instruction
+
+
+@dataclass
+class _MeasureInfo:
+    creg_name: str
+    num_bits: int
+    num_bytes: int
+    qreg_indices: list[int]
+
+
+class Sampler(BaseSamplerV2):
+    """
+    Simple implementation of :class:`BaseSamplerV2` with Statevector.
+    """
+
+    _DEFAULT_SHOTS: int = 512
+
+    def __init__(self, *, seed: np.random.Generator | int | None = None):
+        """
+        Args:
+            seed: The seed for random number generator.
+        """
+        self._seed = seed
+        if isinstance(self._seed, np.random.Generator):
+            self._rng = self._seed
+        else:
+            self._rng = np.random.default_rng(self._seed)
+
+    @property
+    def seed(self) -> np.random.Generator | int | None:
+        """Return the seed for random number generator.
+
+        Returns:
+            np.random.Generator | int | None: The seed for random number generator.
+        """
+        return self._seed
+
+    def run(
+        self, pubs: Iterable[SamplerPubLike], *, shots: int | None = None
+    ) -> PrimitiveJob[PrimitiveResult[PubResult]]:
+        job: PrimitiveJob[PubResult] = PrimitiveJob(self._run, pubs, shots)
+        job._submit()
+        return job
+
+    def _run(
+        self, pubs: Iterable[SamplerPubLike], shots: int | None = None
+    ) -> PrimitiveResult[PrimitiveResult[PubResult]]:
+        if shots is None:
+            shots = self._DEFAULT_SHOTS
+        coerced_pubs = [SamplerPub.coerce(pub, shots) for pub in pubs]
+        for pub in coerced_pubs:
+            pub.validate()
```

Comment:
Since `.coerce` does validation on non-pub inputs, and the pub constructor does validation by default, I think the following would be better to avoid double validation:

```suggestion
        coerced_pubs = [SamplerPub.coerce(pub, shots) for pub in pubs]
```

In this case, user would have to go out of their way to provide an invalid pub.